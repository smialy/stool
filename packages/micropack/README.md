# smicro
## Simple bundler for tiny modules, powered by [rollup](https://github.com/rollup/rollup)


## Installation

`npm i -D @stool/micropack`

Example of `package.json`:

```json
{
"source": "src/index.(js|mjs|jsx|ts|tsx)",
"main": "dist/foo.js",
"module": "dist/foo.mjs",
"types": "dist/types/index.d.ts",
"scripts": {
    "build": "micropack",
    "dev": "micropack --watch"
}
```

For multiple entries: `micropack.json`
```json
[{
    "input": "src/bundler.mjs",
    "output": ["dist/smicro.js", "dist/smicro.mjs"]
}, {
    "input": "src/cli.mjs",
    "output": [{
        "file":"dist/cli.mjs",
        "cli": true
    }]
}]
```

## License

MIT
