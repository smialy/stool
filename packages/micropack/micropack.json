[
    {
        "input": "src/micropack.mjs",
        "output": [
            "dist/micropack.js",
            "dist/micropack.mjs"
        ]
    },
    {
        "input": "src/cli.mjs",
        "output": [
            {
                "file": "dist/cli.js",
                "cli": true
            }
        ]
    }
]