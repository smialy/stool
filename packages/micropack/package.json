{
  "name": "@stool/micropack",
  "version": "0.3.0",
  "description": "Simple bundler for tiny modules powered by rollup",
  "type": "module",
  "bin": {
    "micropack": "src/cli.mjs"
  },
  "exports": {
    ".": {
      "require": "./dist/micropack.js",
      "import": "./src/micropack.mjs"
    }
  },
  "scripts": {
    "format": "prettier --write ./src",
    "lint": "eslint src/**/*.mjs",
    "test": "cross-env NODE_OPTIONS=--experimental-vm-modules jest test/index.test.js"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/smialy/stool/tree/master/packages/micropack"
  },
  "jest": {
    "testEnvironment": "node"
  },
  "license": "MIT",
  "dependencies": {
    "@rollup/plugin-json": "4.1.0",
    "@rollup/plugin-node-resolve": "13.0.0",
    "@swc/core": "1.2.61",
    "commander": "8.0.0",
    "rollup": "2.52.3",
    "rollup-plugin-terser": "^7.0.2",
    "rollup-plugin-typescript2": "^0.30.0",
    "typescript": "4.3.4"
  },
  "files": [
    "src/**/*.*",
    "README.md"
  ],
  "keywords": [
    "bundle",
    "rollup",
    "micro library"
  ],
  "devDependencies": {
    "cross-env": "^7.0.3",
    "eslint": "^7.29.0",
    "jest": "^27.0.6",
    "prettier": "2.3.2"
  }
}
