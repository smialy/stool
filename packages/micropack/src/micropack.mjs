import { resolve, dirname, extname } from 'path';

import { rollup, watch } from 'rollup';
import json from '@rollup/plugin-json';
import nodeResolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import typescript from 'rollup-plugin-typescript2';
// import buble from '@rollup/plugin-buble';
import css from '@modular-css/rollup';
import cleanup from 'rollup-plugin-cleanup';
import terser from 'rollup-plugin-terser';

import { readJsonFile } from './utils/file.mjs';
import { createBabelConfig } from './utils/babel.mjs';


async function loadConfig(pkg, cwd) {
    let config = await readConfigFile(cwd);
    if (!config) {
        config = findPackageEntries(pkg);
    }
    return validateEntries(config);
}

export default async function micropack(options) {
    if (options.dev) {
        console.log(options);
    }
    options.cwd = resolve(process.cwd(), options.cwd);
    options.pkg = await readPackageFile(options.cwd);
    const { include, entries } = await loadConfig(options.pkg, options.cwd);

    const tasks = await Promise.all(entries
        .reduce((acc, entry) => {
            const { input, output } = entry;
            for (const item of output) {
                acc.push({ include, input, output: item });
            }
            return acc;
        }, [])
        .map(prepareRollupConfig(options)));
    console.log({ tasks })
    if (!tasks.length) {
        console.log('Nothing to build');
        process.exit(1);
    }

    if (options.watch) {
        return new Promise((_, reject) => {
            for (const { input, output } of tasks) {
                watch({
                    ...input,
                    output: output,
                    watch: {
                        exclude: 'node_modules/**',
                    },
                }).on('event', (e) => {
                    if (e.code === 'FATAL') {
                        return reject(e.error);
                    } else if (e.code === 'ERROR') {
                        console.warn(e.error);
                    } else if (e.code === 'END') {
                        console.log('Saved.');
                    }
                });
            }
        });
    }
    for (const { input, output } of tasks) {
        let bundle = await rollup(input);
        await bundle.write(output);
    }
    console.log('Builded.');
}

const prepareRollupConfig = (options) => async ({ include, input, output }) => {
    if (options.dev) {
        console.log('Entry:', { input, output });
    }
    const useTypescript =
        extname(input.file) === '.ts' || extname(input.file) === '.tsx';
    const declarationDir = findTypesEntry(options.pkg);
    const format = options.format
        ? options.format
        : extname(output.file) === '.mjs'
        ? 'es'
        : 'cjs';
    return {
        input: {
            input: resolve(options.cwd, input.file),
            onwarn(warning) {
                console.log({ warning });
            },
            external(id) {
                if (id === 'tslib') {
                    return false;
                }
                if (include.includes(id)) {
                    return false;
                }
            },
            plugins: [
                nodeResolve({
                    mainFields: ['module', 'jsnext', 'main'],
                    browser: false,
                    extensions: ['.mjs', '.js', '.jsx', '.json'],
                    preferBuiltins: false,
                }),
                commonjs({
                    include: /\/node_modules\//,
                }),
                json(),
                css({
                    json: true,
                    meta: true,
                    namedExports: true,
                    styleExport: true,
                    dev: options.dev,
                    verbose: options.dev,
                    empties: true,
                    // namer: shornames(),
                }),
                useTypescript &&
                    typescript({
                        // typescript: typescriptModule,
                        cacheRoot: `./node_modules/.cache/bundle_cache_${format}`,
                        useTsconfigDeclarationDir: true,
                        tsconfigDefaults: {
                            compilerOptions: {
                                sourceMap: options.sourcemap,
                                declaration: !!declarationDir,
                                declarationDir,
                                jsx: 'react',
                                jsxFactory: options.jsx || 'h',
                            },
                        },
                        tsconfig: options.tsconfig,
                        tsconfigOverride: {
                            compilerOptions: {
                                target: 'esnext',
                            },
                        },
                    }),
                !useTypescript && !options.modern && await createBabelConfig(options),
                options.compress &&
                    terser.terser({
                        warnings: true,
                        ecma: options.modern ? 9 : 5,
                        toplevel:
                            options.modern ||
                            format === 'cjs' ||
                            format === 'es',
                    }),
                cleanup({
                    comments: 'none',
                }),
            ],
        },
        output: {
            format,
            banner() {
                if (output.cli) {
                    return '#!/usr/bin/env node\n';
                }
                return `// ${new Date().toISOString()}`;
            },
            name: options.name,
            sourcemap: options.sourcemap,
            strict: options.strict === true,
            file: resolve(options.cwd, output.file),
        },
    };
};

async function readPackageFile(cwd) {
    return await readJsonFile(resolve(cwd, 'package.json'));
}

async function readConfigFile(cwd, configFile = 'micropack.json') {
    try {
        return await readJsonFile(resolve(cwd, configFile));
    } catch (e) {}
}

async function validateEntries(config) {
    if (Array.isArray(config)) {
        config = { include: [], entries: config };
    }
    const entries = [];
    for (let { input, output } of config.entries) {
        if (!Array.isArray(output)) {
            output = [output];
        }
        if (typeof input === 'string') {
            input = {
                file: input,
            };
        }
        output = output.map((file) => {
            if (typeof file === 'string') {
                return {
                    file,
                };
            }
            return file;
        });
        entries.push({
            input,
            output,
        });
    }
    return { include: [], ...config, entries };
}

function findPackageEntries(pkg) {
    const entries = [];
    if (pkg.source) {
        for (const name of ['main', 'module']) {
            if (pkg[name]) {
                entries.push({
                    input: pkg.source,
                    output: {
                        file: pkg[name],
                    },
                });
            }
        }
    } else {
        console.warn('Missing "source" in package.json file.');
    }
    return { entries };
}

function findTypesEntry(pkg) {
    if (pkg.types) {
        return dirname(pkg.types);
    }
}
