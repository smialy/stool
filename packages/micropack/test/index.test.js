import fs from 'fs';
import afs from 'fs/promises';
import * as Path from 'path';
import { fileURLToPath } from 'url';
import micropack from '../src/micropack.mjs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = Path.dirname(__filename);

const FIXTURES_DIR = `${__dirname}/fixtures`;

describe('fixtures', () => {
    const dirs = fs
        .readdirSync(FIXTURES_DIR)
        .filter(fixturePath =>
            fs.statSync(Path.resolve(FIXTURES_DIR, fixturePath)).isDirectory(),
        );
    it.each(dirs)("build package: %s", async dir => {
        const path = Path.resolve(FIXTURES_DIR, dir);
        const stdout = await build(path);
        expect(stdout).toMatchSnapshot();

        const pkg = await readPackeageFile(path);
        const outputFiles = findOutputFiles(pkg);
        const result = await Promise.all(
            outputFiles.map(file => Path.resolve(path, file))
            .map(isFileExist));
        const fileExists = outputFiles.map((file, i) => [file, result[i]]);
        expect(fileExists).toMatchSnapshot();
    });
});

async function build(cwd) {
    const cl = console.log;
    const cw = console.warn;

    const buff = [];
    console.log = msg => buff.push(`log: ${msg}`);
    console.warn = msg => buff.push(`warn: ${msg}`);
    try {
        await micropack({ cwd });
    } catch(e) {
        buff.push(e.toString());
    } finally {
        console.log = cl;
        console.warn = cw;
    }
    return buff.join("\n");
}

async function readPackeageFile(dir) {
    return readJsonFile(Path.resolve(dir, 'package.json'));
}

async function readJsonFile(filePath) {
    const data = await afs.readFile(filePath, "utf-8");
    return JSON.parse(data);
}

function findOutputFiles(pkg) {
    return ['main', 'module', 'types'].map(name => pkg[name]).filter(path => path);
}

async function isFileExist(filePath) {
    try {
        return (await afs.stat(filePath)).isFile();
    } catch(e) {

    }
    return false;
}