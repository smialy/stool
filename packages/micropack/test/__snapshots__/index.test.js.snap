// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixtures build package: empty 1`] = `
"warn: Missing key: \\"source\\" in package.json file.
log: Nothing to build."
`;

exports[`fixtures build package: empty 2`] = `Array []`;

exports[`fixtures build package: simple 1`] = `
"log: Building...
log:     dist/index.js
log:     dist/index.mjs
log: Done."
`;

exports[`fixtures build package: simple 2`] = `
Array [
  Array [
    "dist/index.js",
    true,
  ],
  Array [
    "dist/index.mjs",
    true,
  ],
]
`;
